### DISCUSSION TITLE
Is PyQtGraph right for our project?

### DISCUSSION THREAD

**Discussion by andrea-bistacchi** [pyqtgraph/pyqtgraph#2080] [Q&A]
Hello, we are developing a geological modelling application, based on many different libraries. The GUI is in Qt with multiple windows and in some of them we plot 2D maps combining vector graphics and raster images. 

At the moment we use the Matplotlib backend for this, but performance is becoming an issue as the size and number of objects we plot is increasing. We read that PyQtGraph should be faster and we are wondering if it could be the right solution for us.

The 2D plot units are map units, so we tend to have very large numbers as in UTM coordinates, and images must be georeferenced (shifted and scaled in their geographical position, not just in pixel coordinates).

Points in scatterplots should have custom symbols (e.g. STIX fonts) and should be rotated one by one according to an additional field.

We also need to capture coordinates with mouse, e.g. to draw new lines (and we understand you have tools for selection zoom and pan).

Finally, we need to turn objects on and off quickly, to add and remove them, and setting transparency for vectors and images would be a plus.

Thanks in advance for suggestions!

**Comment by sem-geologist** [comment]
why normal GIS applications does not work for you? You know QGIS has this possibility to get minimal app for custom applications using their canvas (with full geo data support). Also you can install Qgis in special conda environment, so you could get some conda-available libraries at versions you want. Pyqtgraph AFAIK and looked into it could not find any support to geographic data. While all your data would be guaranteed to come in same georeferencing system - that would be no problem, but visualising data with different geographic reference systems would require complete implementation of projection framework. (Albeit Qt itself has some affine transformation methods, but geographic systems and transformations inbetween are no more linear... I don't imagine implementing those in python/qt). Implementing `proj` into pyqtgraph in my opinion makes little sense while there are ready solutions like QGIS available.

**Comment by j9ac9k** [comment]
Hi @andrea-bistacchi 

I migrated the issue to a discussion as this feels a more appropriate place for this sort of discussion.

I'll try and break down your post a bit more for clarity purposes.

> At the moment we use the Matplotlib backend for this, but performance is becoming an issue as the size and number of objects we plot is increasing. We read that PyQtGraph should be faster and we are wondering if it could be the right solution for us.

PyQtGraph is nowhere near feature complete as matplotlib, but the areas of the libraries have overlapping capabilities, pyqtgraph surpasses matplotlib in performance.

> The 2D plot units are map units, so we tend to have very large numbers as in UTM coordinates, and images must be georeferenced (shifted and scaled in their geographical position, not just in pixel coordinates).

I don't know much of those kinds of coordinates, we can generally handle pretty wide range of values, in terms of large values, how large are we talking?  An area that pyqtgraph suffers with is non-linear transformations, Qt doesn't give us a good way of manipulating plotting area in a non-linear fashion.

> Points in scatterplots should have custom symbols (e.g. STIX fonts) and should be rotated one by one according to an additional field.

If you look at our scatter plot example, you can see we can render scatter plots with letters from a specific font.

![image](https://user-images.githubusercontent.com/646398/141652549-1eda8eca-15e4-4ece-90db-3ca7942241d1.png)

Here is the relevant section of the code: https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/examples/ScatterPlot.py#L57-L87

> We also need to capture coordinates with mouse, e.g. to draw new lines (and we understand you have tools for selection zoom and pan).

Some version of this is a fairly common use-case for pyqtgraph, I don't think we have built-in functionality to create new lines from the mouse, but it can be added without much difficulty IMO.

> Finally, we need to turn objects on and off quickly, to add and remove them, and setting transparency for vectors and images would be a plus.

I don't see why this wouldn't be an issue off the top of my head.

I don't know what the needs are regarding converting one coordinate system to another  via non-linear transformations, but that's currently a pain point in this library.  It could be made to work, but the code complexity is substantial and there would be a performance hit.  For linear transformations (scaling, rotating, shearing) that's no problem.

Generally would advise looking at our example app `python -m pyqtgraph.examples` and take a look at what you think might be relevant examples.  If one example has similar functionality to what you're looking for, I would recommend tweaking it to try and achieve the desired outcome, or if you need help open an issue/discussion or post on the mail list asking for some assistance on how to get a particular bit of functionality.

Hope that helps!

**Reply by andrea-bistacchi** [reply]
Hi, thanks for the encouraging answer!

> Hi @andrea-bistacchi
> 
> I migrated the issue to a discussion as this feels a more appropriate place for this sort of discussion.
> 

Well done thanks!

> I'll try and break down your post a bit more for clarity purposes.
> 
> > At the moment we use the Matplotlib backend for this, but performance is becoming an issue as the size and number of objects we plot is increasing. We read that PyQtGraph should be faster and we are wondering if it could be the right solution for us.
> 
> PyQtGraph is nowhere near feature complete as matplotlib, but the areas of the libraries have overlapping capabilities, pyqtgraph surpasses matplotlib in performance.
> 

This in what we are looking for!

> > The 2D plot units are map units, so we tend to have very large numbers as in UTM coordinates, and images must be georeferenced (shifted and scaled in their geographical position, not just in pixel coordinates).
> 
> I don't know much of those kinds of coordinates, we can generally handle pretty wide range of values, in terms of large values, how large are we talking? An area that pyqtgraph suffers with is non-linear transformations, Qt doesn't give us a good way of manipulating plotting area in a non-linear fashion.
> 

Coordinates are linear and no transformation is required. Basically they can be numbers in metres like x = 645830.321 y = 5234789.123 z = 2345.123, where x is the east coordinate, y is north and z is elevation above sea level. The only problem is we need precision in millimetres, sometimes even less, keeping the large numbers of UTM (universal transverse Mercator) cords.

> > Points in scatterplots should have custom symbols (e.g. STIX fonts) and should be rotated one by one according to an additional field.
> 
> If you look at our scatter plot example, you can see we can render scatter plots with letters from a specific font.
> 
> ![image](https://user-images.githubusercontent.com/646398/141652549-1eda8eca-15e4-4ece-90db-3ca7942241d1.png)
> 
> Here is the relevant section of the code: https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/examples/ScatterPlot.py#L57-L87
> 

Looks good, thanks! STIX fonts would be ok?

> > We also need to capture coordinates with mouse, e.g. to draw new lines (and we understand you have tools for selection zoom and pan).
> 
> Some version of this is a fairly common use-case for pyqtgraph, I don't think we have built-in functionality to create new lines from the mouse, but it can be added without much difficulty IMO.
> 

We already have functions for that. We just need to pick xy coordinates with mouse clicks  in the axes reference while possibly keeping zooming and panning active, and the Qt GUI active “out” of the figure canvas (this is achieved with different event loops when using the Matplotlib Qt backend in a Qt application).

> > Finally, we need to turn objects on and off quickly, to add and remove them, and setting transparency for vectors and images would be a plus.
> 
> I don't see why this wouldn't be an issue off the top of my head.
> 

Do you have a system of “actors” or “artists” as in Matplotlib that get a reference when they are created and then can be turned on or off, and then completely destroyed if no more needed?

> I don't know what the needs are regarding converting one coordinate system to another via non-linear transformations, but that's currently a pain point in this library. It could be made to work, but the code complexity is substantial and there would be a performance hit. For linear transformations (scaling, rotating, shearing) that's no problem.
> 

This is not a problem as said in the other answer to @sem-geologist.

> Generally would advise looking at our example app `python -m pyqtgraph.examples` and take a look at what you think might be relevant examples. If one example has similar functionality to what you're looking for, I would recommend tweaking it to try and achieve the desired outcome, or if you need help open an issue/discussion or post on the mail list asking for some assistance on how to get a particular bit of functionality.
> 
> Hope that helps!

Thanks very much!

**Reply by j9ac9k** [reply]
> Looks good, thanks! STIX fonts would be ok?

I don't know what STIX fonts are, but if you can load it as a `QFont` instance, then I don't see why it wouldn't work.

> We already have functions for that. We just need to pick xy coordinates with mouse clicks in the axes reference while possibly keeping zooming and panning active, and the Qt GUI active “out” of the figure canvas (this is achieved with different event loops when using the Matplotlib Qt backend in a Qt application).

Nice, that was going to be my suggestion to not depend too much on pyqtgraph specific functionality but use the base Qt functionality instead.  Given you've already implemented this, it will likely transfer over w/o too much difficulty.

> Do you have a system of “actors” or “artists” as in Matplotlib that get a reference when they are created and then can be turned on or off, and then completely destroyed if no more needed?

I'm not that familiar with matplotlib, so I'm reluctant to say, but we generally have [GraphicsScene.removeItem ](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/GraphicsScene/GraphicsScene.py#L401-L405) method where you call the scene, and tell it to remove an item from it.  


If you don't have non-linear transformations, then pyqtgraph is definitely a library you all should take a very close look at.  I would love to expand more use case for geographic examples, I know the [orange3 application](https://github.com/biolab/orange3) uses pyqtgraph to render maps (although truth be told, I'm not sure how they do it, I should probably take a look at their source code and see what they're doing.

We have recently provided easy/fast way of going from numpy arrays of x,y coordinates to QPolygonF objects which can be used to render all sorts of shapes, and draw fill behavior.

**Reply by andrea-bistacchi** [reply]
Hi and thanks again!

To show an image at particular coordinates I just have to use the _pos_, _scale_ or _transform_ parameters in imageShow, correct?

And how it works if I have several superposed "layers"? Is there a way to ensure that an image stays on the background and for instance lines are drawn in front?

Thanks!

**Reply by sem-geologist** [reply]
From my experience with plotting multiple images and pionts and lines.... I did that all in PlotWidget (particularly that position/coordinates are important). As for a images I was creating `imageItem` and as I had no need for skew, but just rotate, scale (x,y,) and move, so I was fitting the image to `QRect`, which position, z order and transformations I find much easier to change. I could be wrong...

**Comment by andrea-bistacchi** [comment]
Hi and thanks for suggestions, however this is not a GIS application. Actually it is a 3D modelling app and we manage and model data - all in a coherent CRS - as VTK objects and/or Numpy arrays (the two are interoperable). Then for some reason that is long to explain here we also need 2D maps and 2D cross-sections. This is what we do with Matplotlib at the moment, but we are not happy with performance. Maybe the point about “georeferencing” was misleading, but all we need is to show also background images with coordinates that are not the simple image pixel coordinates, as you do in Matplotlib with the “extent” parameter.

**Reply by andrea-bistacchi** [reply]
Thanks!
